File: puppeteer-python/climanuvem/pages/register_page.py
# -*- coding: utf-8 -*-
"""Page Object for the Register form. Constructing this object waits until
the username input is visible. Mirrors selenium-java's ``RegisterPage``.
"""
from climanuvem.pages.base_page import BasePage

REGISTER_FEEDBACK_KEYWORDS = (
    "error",
    "inválid",
    "invalid",
    "contraseña",
    "correo",
    "usuario",
    "coincid",
    "uso",
    "obligatorio",
    "requerid",
)
VERIFY_DIALOG_KEYWORDS = ("verific", "correo")


class RegisterPage(BasePage):
    USERNAME_INPUT = BasePage.input_by_placeholder("Nombre de usuario")
    EMAIL_INPUT = BasePage.input_by_placeholder("Correo electrónico")
    PASSWORD_INPUT = BasePage.input_by_placeholder("Contraseña")
    CONFIRM_PASSWORD_INPUT = BasePage.input_by_placeholder("Confirmar contraseña")
    LOGIN_LINK = BasePage.by_partial_text("Inicia sesión")
    SUBMIT_BUTTON = BasePage.by_xpath(
        "//*[@role='button' or self::button or @tabindex]"
        "[contains(translate(normalize-space(.),"
        "'ABCDEFGHIJKLMNOPQRSTUVWXYZÁÉÍÓÚÜÑ','abcdefghijklmnopqrstuvwxyzáéíóúüñ'),'registr')"
        " or contains(translate(normalize-space(.),"
        "'ABCDEFGHIJKLMNOPQRSTUVWXYZÁÉÍÓÚÜÑ','abcdefghijklmnopqrstuvwxyzáéíóúüñ'),'crear')]"
    )
    HOME_MARKER = BasePage.by_partial_text("Bienvenido")

    def __init__(self, session):
        super().__init__(session)
        self.wait_for(self.USERNAME_INPUT)

    # ── Queries ──────────────────────────────────────────────────────────

    def is_username_input_present(self) -> bool:
        return self.is_present(self.USERNAME_INPUT)

    def is_email_input_present(self) -> bool:
        return self.is_present(self.EMAIL_INPUT)

    def is_password_input_present(self) -> bool:
        return self.is_present(self.PASSWORD_INPUT)

    def is_confirm_password_input_present(self) -> bool:
        return self.is_present(self.CONFIRM_PASSWORD_INPUT)

    def is_login_link_present(self) -> bool:
        return self.is_present(self.LOGIN_LINK)

    def is_home_visible(self) -> bool:
        return self.is_present(self.HOME_MARKER)

    def is_verification_dialog_visible(self) -> bool:
        return self.body_text_contains_any(VERIFY_DIALOG_KEYWORDS[:1]) and self.body_text_contains_any(
            VERIFY_DIALOG_KEYWORDS[1:]
        )

    def has_register_error_or_validation(self) -> bool:
        return self.body_text_contains_any(REGISTER_FEEDBACK_KEYWORDS) or self.has_invalid_required_input()

    # ── Actions ──────────────────────────────────────────────────────────

    def enter_username(self, username: str) -> "RegisterPage":
        self.fill(self.USERNAME_INPUT, username)
        return self

    def enter_email(self, email: str) -> "RegisterPage":
        self.fill(self.EMAIL_INPUT, email)
        return self

    def enter_password(self, password: str) -> "RegisterPage":
        self.fill(self.PASSWORD_INPUT, password)
        return self

    def enter_confirm_password(self, password: str) -> "RegisterPage":
        self.fill(self.CONFIRM_PASSWORD_INPUT, password)
        return self

    def submit_register(self) -> "RegisterPage":
        self.click_last_visible(self.SUBMIT_BUTTON)
        return self

    def register(self, username: str, email: str, password: str, confirm_password: str) -> "RegisterPage":
        self.enter_username(username)
        self.enter_email(email)
        self.enter_password(password)
        self.enter_confirm_password(confirm_password)
        return self.submit_register()

    def wait_for_home(self):
        from climanuvem.pages.home_page import HomePage

        self.wait_for(self.HOME_MARKER)
        return HomePage(self._session)

    def wait_for_verification_dialog(self) -> "RegisterPage":
        predicate = (
            "() => {"
            "const lower = document.body.innerText.toLowerCase();"
            "return lower.includes('verific') && lower.includes('correo');"
            "}"
        )
        self._run(self._page.waitForFunction(predicate, {"timeout": 30_000}))
        return self

    def wait_for_register_failure(self) -> "RegisterPage":
        keywords_js = "[" + ",".join(f"'{keyword}'" for keyword in REGISTER_FEEDBACK_KEYWORDS) + "]"
        predicate = (
            "() => {"
            "const lower = document.body.innerText.toLowerCase();"
            f"const feedback = {keywords_js}.some(k => lower.includes(k));"
            "const invalid = Array.from(document.querySelectorAll('input'))"
            ".some(el => el.required && !el.checkValidity());"
            f"const usernamePresent = !!document.querySelector('{self.USERNAME_INPUT.value}');"
            "return (feedback || invalid) && usernamePresent;"
            "}"
        )
        self._run(self._page.waitForFunction(predicate, {"timeout": 30_000}))
        return self

    def click_login_link(self):
        """Navigates back to the Login form."""
        from climanuvem.pages.login_page import LoginPage

        self.click(self.LOGIN_LINK)
        return LoginPage(self._session)

File: cypress-js/cypress/support/pages/RegisterPage.js
import BasePage from './BasePage';
import LoginPage from './LoginPage';
import HomePage from './HomePage';
import { bodyTextContainsAny, hasInvalidRequiredInput } from '../common/domChecks';

const USERNAME_PLACEHOLDER = 'Nombre de usuario';
const EMAIL_PLACEHOLDER = 'Correo electrónico';
const PASSWORD_PLACEHOLDER = 'Contraseña';
const CONFIRM_PASSWORD_PLACEHOLDER = 'Confirmar contraseña';
const LOGIN_LINK_TEXT = 'Inicia sesión';
const HOME_MARKER_TEXT = 'Bienvenido';
const INTERACTIVE_SELECTOR = '[role="button"], button, [tabindex]';
const SUBMIT_KEYWORDS = ['registr', 'crear'];
const VERIFY_DIALOG_KEYWORDS = ['verific', 'correo'];
const REGISTER_FEEDBACK_KEYWORDS = [
  'error',
  'inválid',
  'invalid',
  'contraseña',
  'correo',
  'usuario',
  'coincid',
  'uso',
  'obligatorio',
  'requerid',
];

/**
 * Page Object for the Register form. Constructing this object waits until
 * the username input is visible.
 */
export default class RegisterPage extends BasePage {
  constructor() {
    super();
    this.byPlaceholder(USERNAME_PLACEHOLDER).should('be.visible');
  }

  // ── Assertions / queries ─────────────────────────────────────────────────

  assertUsernameInputVisible() {
    this.byPlaceholder(USERNAME_PLACEHOLDER).should('be.visible');
    return this;
  }

  assertEmailInputVisible() {
    this.byPlaceholder(EMAIL_PLACEHOLDER).should('be.visible');
    return this;
  }

  assertPasswordInputVisible() {
    this.byPlaceholder(PASSWORD_PLACEHOLDER).should('be.visible');
    return this;
  }

  assertConfirmPasswordInputVisible() {
    this.byPlaceholder(CONFIRM_PASSWORD_PLACEHOLDER).should('be.visible');
    return this;
  }

  assertLoginLinkVisible() {
    this.byPartialText(LOGIN_LINK_TEXT).should('be.visible');
    return this;
  }

  /** Yields (Cypress-chainable) true when the "verify your email" guidance is showing. */
  isVerificationDialogVisible() {
    return cy.get('body').then(($body) => {
      const text = $body.text().toLowerCase();
      return VERIFY_DIALOG_KEYWORDS.every((keyword) => text.includes(keyword));
    });
  }

  /** Yields (Cypress-chainable) true when a register error or validation feedback is showing. */
  hasRegisterErrorOrValidation() {
    return cy.get('body').then(($body) => {
      return bodyTextContainsAny($body, REGISTER_FEEDBACK_KEYWORDS) || hasInvalidRequiredInput($body[0].ownerDocument);
    });
  }

  // ── Actions ───────────────────────────────────────────────────────────────

  enterUsername(username) {
    this.byPlaceholder(USERNAME_PLACEHOLDER).clear({ force: true }).type(username, { force: true });
    return this;
  }

  enterEmail(email) {
    this.byPlaceholder(EMAIL_PLACEHOLDER).clear({ force: true }).type(email, { force: true });
    return this;
  }

  enterPassword(password) {
    this.byPlaceholder(PASSWORD_PLACEHOLDER).clear({ force: true }).type(password, { force: true });
    return this;
  }

  enterConfirmPassword(password) {
    this.byPlaceholder(CONFIRM_PASSWORD_PLACEHOLDER).clear({ force: true }).type(password, { force: true });
    return this;
  }

  submitRegister() {
    cy.get('body').then(($body) => {
      const candidates = Cypress.$.makeArray($body.find(INTERACTIVE_SELECTOR)).filter((el) => {
        const text = el.textContent.trim().toLowerCase();
        return SUBMIT_KEYWORDS.some((keyword) => text.includes(keyword));
      });
      const visible = [...candidates].reverse().find((el) => Cypress.$(el).is(':visible'));
      cy.wrap(visible).click({ force: true });
    });
    return this;
  }

  register(username, email, password, confirmPassword) {
    this.enterUsername(username);
    this.enterEmail(email);
    this.enterPassword(password);
    this.enterConfirmPassword(confirmPassword);
    return this.submitRegister();
  }

  waitForHome() {
    this.byPartialText(HOME_MARKER_TEXT).should('be.visible');
    return new HomePage();
  }

  waitForVerificationDialog() {
    cy.get('body', { timeout: 30000 }).should(($body) => {
      const text = $body.text().toLowerCase();
      const shown = VERIFY_DIALOG_KEYWORDS.every((keyword) => text.includes(keyword));
      if (!shown) {
        throw new Error('Verification dialog did not appear before timeout');
      }
    });
    return this;
  }

  waitForRegisterFailure() {
    cy.get('body', { timeout: 30000 }).should(($body) => {
      const hasFailure =
        bodyTextContainsAny($body, REGISTER_FEEDBACK_KEYWORDS) || hasInvalidRequiredInput($body[0].ownerDocument);
      const usernamePresent = $body.find(`input[placeholder="${USERNAME_PLACEHOLDER}"]`).length > 0;
      if (!(hasFailure && usernamePresent)) {
        throw new Error('Registration was not rejected before timeout');
      }
    });
    return this;
  }

  /** Navigates back to the Login form. */
  clickLoginLink() {
    this.clickByText(LOGIN_LINK_TEXT);
    return new LoginPage();
  }
}
